dev/service-registry/service-locator/src/main/scala/com/lightbend/lagom/registry/impl/ServiceRegistryActor.scala

```diff
@@ -19,100 +19,182 @@ import scala.collection.JavaConverters._
   case class Remove(name: String)
   case class Register(name: String, service: ServiceRegistryService)
-  case class Route(method: String, path: String)
+  case class Route(method: String, path: String, portName: Option[String])
   case object GetRegisteredServices
   case class RegisteredServices(services: PSequence[RegisteredService])
   sealed trait RouteResult
-  case class Found(address: InetSocketAddress) extends RouteResult
-  case class NotFound(registry: Map[String, ServiceRegistryService]) extends RouteResult
+  case class Found(address: URI) extends RouteResult
+  case class NotFound(registry: Map[String, (URI, ServiceRegistryService)]) extends RouteResult
+
 }
 
-class ServiceRegistryActor @Inject() (unmanagedServices: UnmanagedServices) extends Actor {
+case class ServiceName(name: String) extends AnyVal
+
+// identifies a single entry on the registry. This Registry will expand portname's `Some(tcp)`
+// and `Some(`http`)` duplicating the entry on the registry to also use `None` for backwards compatibility
+case class ServiceRegistryKey(serviceName: ServiceName, portName: Option[String])
+
+object InternalRegistry {
+
+  def build(unmanagedServices: UnmanagedServices): InternalRegistry = new InternalRegistry(
+    unmanagedServices.services.flatMap {
+      case (serviceName, serviceRegistryService) =>
+        serviceRegistryService.uris().asScala.flatMap {
+          buildRegistryItem(serviceName, serviceRegistryService)
+        }
+    }.toMap
+  )
+
+  private def buildRegistryItem(serviceName: String, serviceRegistryService: ServiceRegistryService)(serviceUri: URI): Seq[(ServiceRegistryKey, (URI, ServiceRegistryService))] = {
+    val portNames = {
+      serviceUri.getScheme match {
+        case "tcp" =>
+          Seq(None) // using "tcp://" defaults to having no portName
+        case "http" =>
+          // using "http://" defaults to having no portName and also Some("http") so "http" becomes the
+          // default result when searching without a `portName` query.
+          Seq(None, Some(serviceUri.getScheme))
+        case _ => Seq(Some(serviceUri.getScheme))
+      }
+    }
+    val registryItems: Seq[(ServiceRegistryKey, (URI, ServiceRegistryService))] = portNames.map { pn =>
+      val srk = ServiceRegistryKey(ServiceName(serviceName), pn)
+      srk -> (serviceUri, serviceRegistryService)
+    }
+    registryItems
+  }
+
+  def build(serviceName: String, details: ServiceRegistryService): Map[ServiceRegistryKey, (URI, ServiceRegistryService)] =
+    details.uris().asScala.flatMap {
+      buildRegistryItem(serviceName, details)
+    }.toMap
 
+}
+
+/**
+ * @param reg map using servicename and portname as keys and a single URI as value. The original ServiceRegistryService
+ *            where eack K/V in the map was extracted from is added into the value to be able to traceback.
+ */
+class InternalRegistry(var reg: Map[ServiceRegistryKey, (URI, ServiceRegistryService)]) {
```

Comment:
It would be better to have a model to use as value here instead of a tuple. This is causing the usage of `._1` and `._2` later and we are loosing context.